### ISSUE TITLE
Type annotations / type hints / stub files for mypy

### ISSUE THREAD

**Issue by MartinThoma** [dynaconf/dynaconf#448]
Are there plans to add type annotations or stub files for mypy?

I've written an [article about type annotations](https://medium.com/analytics-vidhya/type-annotations-in-python-3-8-3b401384403d) if you need one :-)

**Comment by rochacbruno** [comment]
@MartinThoma pull requests adding annotations and stubs are very welcome.

**Comment by MartinThoma** [comment]
I see that you're supporting Python 3.6+. Then I would prefer type annotations instead of stub files. I'll prepare something once I have some time :-)

**Comment by MartinThoma** [comment]
https://github.com/rochacbruno/dynaconf/pull/450

**Comment by disrupted** [comment]
commenting because this is still a valid and desired request, therefore shouldn't be closed

**Comment by SonGokussj4** [comment]
Still valid even year later. Any progress? I tried to implement some dataclasses with pydantic but faild. Does anyone have a working workaround?

**Comment by MartinThoma** [comment]
I no longer use dynaconf

**Comment by FeryET** [comment]
Any update on this?

**Comment by apowers313** [comment]
Here was my workaround: convert the Dynaconf config into a Pydantic BaseModel. Pydantic does all the validation and defaults and creates a strongly-typed config.

The code is a bit long (but conceptually very simple) because it has to account for all the default Dynaconf fields, so the Gist is here: https://gist.github.com/apowers313/c009991a31195e9e3ee8dc51b989136a

**Comment by LukeSavefrogs** [comment]
Yet another **workaround** inspired by @apowers313 approach:

```python
# -------------- src/package/config/schemas.py -------------- 
import dataclasses
from pathlib import Path

from dynaconf import LazySettings

@dataclasses.dataclass
class SettingsSchema(LazySettings):
    myProperty: str
    myCategory: "CategorySettingsSchema"


@dataclasses.dataclass
class CategorySettingsSchema:
    mySecondProperty: int


# -------------- src/package/config/__init__.py -------------- 
from dynaconf import Dynaconf
from package.config.schemas import SettingsSchema

settings: SettingsSchema = Dynaconf(  # pyright: ignore[reportGeneralTypeIssues]
    ...
)
```

- **PROS**: 
   - Less verbose
   - Does not need any other dependency installed (`dataclasses` is shipped with Python)
   - Only access what you have in the configuration files

- **CONS**:
   - Only access what you have in the configuration files
   - Only supports accessing properties via the **dot operator** (`settings.property.subproperty`)


For simpler configurations like mine could be useful.

**Comment by apowers313** [comment]
Great idea / example @LukeSavefrogs :)

Note that you can change [line 75](https://gist.github.com/apowers313/c009991a31195e9e3ee8dc51b989136a#file-dynaconf_plus_pydantic-py-L75) of my example from `extra="forbid"` to `extra="allow"` and get pretty much the same thing as the dataclass example -- but it won't catch extra fields that have been defined in the config that you weren't expecting (a potential source of errors).

Also, put in the PROS column for Pydantic that it has [more validation types](https://docs.pydantic.dev/latest/usage/types/types/) (like email addresses, colors, IP addresses, URLs, etc.) and stuff like [discriminated unions](https://docs.pydantic.dev/latest/usage/types/unions/#discriminated-unions-aka-tagged-unions) which could enable different validation requirements for dev / prod. (I also just noticed that it has [Settings Management](https://docs.pydantic.dev/latest/usage/pydantic_settings/), which might be worth looking into).

I'm really liking the separation of duties of Dynaconf and Pydantic.

**Comment by LukeSavefrogs** [comment]
> Great idea / example @LukeSavefrogs :)
> 
> Note that you can change [line 75](https://gist.github.com/apowers313/c009991a31195e9e3ee8dc51b989136a#file-dynaconf_plus_pydantic-py-L75) of my example from `extra="forbid"` to `extra="allow"` and get pretty much the same thing as the dataclass example -- but it won't catch extra fields that have been defined in the config that you weren't expecting (a potential source of errors).

Thank you for your clarification, the only reason why i posted this (very dirty) "workaround" is because i **didn't want to install another dependency** that I will use only while developing and will be packaged with the final application increasing the bundle size and start times without any improvement for the user.

> Also, put in the PROS column for Pydantic that it has [more validation types](https://docs.pydantic.dev/latest/usage/types/types/)

Sorry for the confusion, the PROS/CONS section was only for the `dataclass` method, I improved it a bit to make it clearer. 😄


In conclusion, your solution is more complete, but I cannot use it because I cannot afford to install a library for no gain for the end-user, so I had to come up with this 😁

**Comment by rochacbruno** [comment]
Fixed on #1107 and being improved tracked on https://github.com/dynaconf/dynaconf/issues/1082#issuecomment-2199703596
